corehq/motech/generic_inbound/utils.py

```diff
@@ -63,58 +65,56 @@ def from_request(cls, request):
         except UnicodeDecodeError:
             raise GenericInboundUserError(_("Unable to decode request body"))
 
-        try:
-            request_json = json.loads(body)
-        except json.JSONDecodeError:
-            raise GenericInboundUserError(_("Payload must be valid JSON"))
         return cls(
             domain=request.domain,
             couch_user=request.couch_user,
             request_method=request.method,
             user_agent=request.META.get('HTTP_USER_AGENT'),
-            data=request_json,
+            data=body,
             query=dict(request.GET.lists()),
-            headers=get_headers_for_api_context(request)
+            headers=get_headers_for_api_context(request),
+            request_id=request_id or uuid.uuid4().hex,
         )
 
     @classmethod
     def from_log(cls, log):
-        try:
-            request_json = json.loads(log.request_body)
-        except (UnicodeDecodeError, json.JSONDecodeError):
-            raise GenericInboundUserError(_("Payload must be valid JSON"))
         return cls(
             domain=log.domain,
             couch_user=CouchUser.get_by_username(log.username),
             request_method=log.request_method,
             user_agent=log.request_headers.get('HTTP_USER_AGENT'),
-            data=request_json,
+            data=log.request_body,
             query=dict(QueryDict(log.request_query).lists()),
             headers=dict(log.request_headers),
+            request_id=log.id
         )
 
-    def to_context(self):
-        restore_user = self.couch_user.to_ota_restore_user(
+    @cached_property
+    def restore_user(self):
+        return self.couch_user.to_ota_restore_user(
             self.domain, request_user=self.couch_user)
-        return get_evaluation_context(
-            restore_user,
-            self.request_method,
-            self.query,
-            self.headers,
-            self.data,
-        )
 
 
 @attr.s(kw_only=True, frozen=True, auto_attribs=True)
 class ApiResponse:
+    """Data class for managing response data and producing HTTP responses.
+    Override ``_get_http_response`` to return different HTTP response."""
     status: int
-    data: dict = None
+    internal_response: dict = None
```

Comment:
this makes sense - I was wondering how you'd handle the different response formats

Reply:
I've added an 'external_response' attribute now that is just a string representing the encoded response data.